### ISSUE TITLE
Gradient Tape (tf.GradientTape) Returning All 0 Values in GradCam

### ISSUE THREAD

**Issue by varungupta31** [tensorflow/tensorflow#56657]
### System information

-   **Have I written custom code (as opposed to using a stock example script
    provided in TensorFlow)**: Yes (Custom Loss + GradCam)
-   **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: CentOS Linux 7
-   **TensorFlow installed from (source or binary)**: PIP
-   **TensorFlow version (use command below)**: 2.9.0
-   **Python version**: 3.7.0
-   **CUDA/cuDNN version**:  11.6
-   **GPU model and memory**: Nvidia RTX 2080Ti

### Describe the problem

I have trained 2 FCN segmentation models, with the same custom loss.  
Model 1 ------>  Trained of 256x512 images, on TF==1.15. 
Model 2 ------> Trained on 512x1024 images, on TF==2.9.0.

**The same GradCam Script works fine on `Model 1`, but returns 0 (and eventually NaN) for `Model 2`.**

**Model1.summary():**

```
 Layer (type)                   Output Shape         Param #     Connected to                     
==================================================================================================
 input_1 (InputLayer)           [(None, 256, 512, 3  0           []                               
                                )]                                                                
                                                                                                  
 block1_conv2 (Conv2D)          (None, 256, 512, 64  1792        ['input_1[0][0]']                
                                )                                                                 
                                                                                                  
 block1_pool (MaxPooling2D)     (None, 128, 256, 64  0           ['block1_conv2[0][0]']           
                                )                                                                 
                                                                                                  
 block2_conv1 (Conv2D)          (None, 128, 256, 12  73856       ['block1_pool[0][0]']            
                                8)                                                                
                                                                                                  
 block2_conv2 (Conv2D)          (None, 128, 256, 12  147584      ['block2_conv1[0][0]']           
                                8)                                                                
                                                                                                  
 block2_pool (MaxPooling2D)     (None, 64, 128, 128  0           ['block2_conv2[0][0]']           
                                )                                                                 
                                                                                                  
 block3_conv1 (Conv2D)          (None, 64, 128, 256  295168      ['block2_pool[0][0]']            
                                )                                                                 
                                                                                                  
 block3_conv2 (Conv2D)          (None, 64, 128, 256  590080      ['block3_conv1[0][0]']           
                                )                                                                 
                                                                                                  
 block3_conv3 (Conv2D)          (None, 64, 128, 256  590080      ['block3_conv2[0][0]']           
                                )                                                                 
                                                                                                  
 block3_conv4 (Conv2D)          (None, 64, 128, 256  590080      ['block3_conv3[0][0]']           
                                )                                                                 
                                                                                                  
 block3_pool (MaxPooling2D)     (None, 32, 64, 256)  0           ['block3_conv4[0][0]']           
                                                                                                  
 block4_conv1 (Conv2D)          (None, 32, 64, 512)  1180160     ['block3_pool[0][0]']            
                                                                                                  
 block4_conv2 (Conv2D)          (None, 32, 64, 512)  2359808     ['block4_conv1[0][0]']           
                                                                                                  
 block4_conv3 (Conv2D)          (None, 32, 64, 512)  2359808     ['block4_conv2[0][0]']           
                                                                                                  
 block4_conv4 (Conv2D)          (None, 32, 64, 512)  2359808     ['block4_conv3[0][0]']           
                                                                                                  
 block4_pool (MaxPooling2D)     (None, 16, 32, 512)  0           ['block4_conv4[0][0]']           
                                                                                                  
 block5_conv1 (Conv2D)          (None, 16, 32, 512)  2359808     ['block4_pool[0][0]']            
                                                                                                  
 block5_conv2 (Conv2D)          (None, 16, 32, 512)  2359808     ['block5_conv1[0][0]']           
                                                                                                  
 block5_conv3 (Conv2D)          (None, 16, 32, 512)  2359808     ['block5_conv2[0][0]']           
                                                                                                  
 block5_conv4 (Conv2D)          (None, 16, 32, 512)  2359808     ['block5_conv3[0][0]']           
                                                                                                  
 block5_pool (MaxPooling2D)     (None, 8, 16, 512)   0           ['block5_conv4[0][0]']           
                                                                                                  
 conv2d (Conv2D)                (None, 8, 16, 4096)  102764544   ['block5_pool[0][0]']            
                                                                                                  
 re_lu (ReLU)                   (None, 8, 16, 4096)  0           ['conv2d[0][0]']                 
                                                                                                  
 dropout (Dropout)              (None, 8, 16, 4096)  0           ['re_lu[0][0]']                  
                                                                                                  
 conv2d_1 (Conv2D)              (None, 8, 16, 4096)  16781312    ['dropout[0][0]']                
                                                                                                  
 re_lu_1 (ReLU)                 (None, 8, 16, 4096)  0           ['conv2d_1[0][0]']               
                                                                                                  
 dropout_1 (Dropout)            (None, 8, 16, 4096)  0           ['re_lu_1[0][0]']                
                                                                                                  
 conv2d_2 (Conv2D)              (None, 8, 16, 2)     8194        ['dropout_1[0][0]']              
                                                                                                  
 conv2d_transpose (Conv2DTransp  (None, 16, 32, 2)   66          ['conv2d_2[0][0]']               
 ose)                                                                                             
                                                                                                  
 conv2d_3 (Conv2D)              (None, 16, 32, 2)    1026        ['block4_pool[0][0]']            
                                                                                                  
 add (Add)                      (None, 16, 32, 2)    0           ['conv2d_transpose[0][0]',       
                                                                  'conv2d_3[0][0]']               
                                                                                                  
 conv2d_transpose_1 (Conv2DTran  (None, 32, 64, 2)   66          ['add[0][0]']                    
 spose)                                                                                           
                                                                                                  
 conv2d_4 (Conv2D)              (None, 32, 64, 2)    514         ['block3_pool[0][0]']            
                                                                                                  
 add_1 (Add)                    (None, 32, 64, 2)    0           ['conv2d_transpose_1[0][0]',     
                                                                  'conv2d_4[0][0]']               
                                                                                                  
 conv2d_transpose_2 (Conv2DTran  (None, 256, 512, 2)  1026       ['add_1[0][0]']                  
 spose)                                                                                           
                                                                                                  
==================================================================================================
Total params: 139,544,204
Trainable params: 139,544,204
Non-trainable params: 0
```

**Model2.summary():**

```
Layer (type)                   Output Shape         Param #     Connected to                     
==================================================================================================
 input_1 (InputLayer)           [(None, 512, 1024,   0           []                               
                                3)]                                                               
                                                                                                  
 block1_conv1 (Conv2D)          (None, 512, 1024, 6  1792        ['input_1[0][0]']                
                                4)                                                                
                                                                                                  
 block1_conv2 (Conv2D)          (None, 512, 1024, 6  36928       ['block1_conv1[0][0]']           
                                4)                                                                
                                                                                                  
 block1_pool (MaxPooling2D)     (None, 256, 512, 64  0           ['block1_conv2[0][0]']           
                                )                                                                 
                                                                                                  
 block2_conv1 (Conv2D)          (None, 256, 512, 12  73856       ['block1_pool[0][0]']            
                                8)                                                                
                                                                                                  
 block2_conv2 (Conv2D)          (None, 256, 512, 12  147584      ['block2_conv1[0][0]']           
                                8)                                                                
                                                                                                  
 block2_pool (MaxPooling2D)     (None, 128, 256, 12  0           ['block2_conv2[0][0]']           
                                8)                                                                
                                                                                                  
 block3_conv1 (Conv2D)          (None, 128, 256, 25  295168      ['block2_pool[0][0]']            
                                6)                                                                
                                                                                                  
 block3_conv2 (Conv2D)          (None, 128, 256, 25  590080      ['block3_conv1[0][0]']           
                                6)                                                                
                                                                                                  
 block3_conv3 (Conv2D)          (None, 128, 256, 25  590080      ['block3_conv2[0][0]']           
                                6)                                                                
                                                                                                  
 block3_conv4 (Conv2D)          (None, 128, 256, 25  590080      ['block3_conv3[0][0]']           
                                6)                                                                
                                                                                                  
 block3_pool (MaxPooling2D)     (None, 64, 128, 256  0           ['block3_conv4[0][0]']           
                                )                                                                 
                                                                                                  
 block4_conv1 (Conv2D)          (None, 64, 128, 512  1180160     ['block3_pool[0][0]']            
                                )                                                                 
                                                                                                  
 block4_conv2 (Conv2D)          (None, 64, 128, 512  2359808     ['block4_conv1[0][0]']           
                                )                                                                 
                                                                                                  
 block4_conv3 (Conv2D)          (None, 64, 128, 512  2359808     ['block4_conv2[0][0]']           
                                )                                                                 
                                                                                                  
 block4_conv4 (Conv2D)          (None, 64, 128, 512  2359808     ['block4_conv3[0][0]']           
                                )                                                                 
                                                                                                  
 block4_pool (MaxPooling2D)     (None, 32, 64, 512)  0           ['block4_conv4[0][0]']           
                                                                                                  
 block5_conv1 (Conv2D)          (None, 32, 64, 512)  2359808     ['block4_pool[0][0]']            
                                                                                                  
 block5_conv2 (Conv2D)          (None, 32, 64, 512)  2359808     ['block5_conv1[0][0]']           
                                                                                                  
 block5_conv3 (Conv2D)          (None, 32, 64, 512)  2359808     ['block5_conv2[0][0]']           
                                                                                                  
 block5_conv4 (Conv2D)          (None, 32, 64, 512)  2359808     ['block5_conv3[0][0]']           
                                                                                                  
 block5_pool (MaxPooling2D)     (None, 16, 32, 512)  0           ['block5_conv4[0][0]']           
                                                                                                  
 conv2d (Conv2D)                (None, 16, 32, 4096  102764544   ['block5_pool[0][0]']            
                                )                                                                 
                                                                                                  
 re_lu (ReLU)                   (None, 16, 32, 4096  0           ['conv2d[0][0]']                 
                                )                                                                 
                                                                                                  
 dropout (Dropout)              (None, 16, 32, 4096  0           ['re_lu[0][0]']                  
                                )                                                                 
                                                                                                  
 conv2d_1 (Conv2D)              (None, 16, 32, 4096  16781312    ['dropout[0][0]']                
                                )                                                                 
                                                                                                  
 re_lu_1 (ReLU)                 (None, 16, 32, 4096  0           ['conv2d_1[0][0]']               
                                )                                                                 
                                                                                                  
 dropout_1 (Dropout)            (None, 16, 32, 4096  0           ['re_lu_1[0][0]']                
                                )                                                                 
                                                                                                  
 conv2d_2 (Conv2D)              (None, 16, 32, 2)    8194        ['dropout_1[0][0]']              
                                                                                                  
 conv2d_transpose (Conv2DTransp  (None, 32, 64, 2)   66          ['conv2d_2[0][0]']               
 ose)                                                                                             
                                                                                                  
 conv2d_3 (Conv2D)              (None, 32, 64, 2)    1026        ['block4_pool[0][0]']            
                                                                                                  
 add (Add)                      (None, 32, 64, 2)    0           ['conv2d_transpose[0][0]',       
                                                                  'conv2d_3[0][0]']               
                                                                                                  
 conv2d_transpose_1 (Conv2DTran  (None, 64, 128, 2)  66          ['add[0][0]']                    
 spose)                                                                                           
                                                                                                  
 conv2d_4 (Conv2D)              (None, 64, 128, 2)   514         ['block3_pool[0][0]']            
                                                                                                  
 add_1 (Add)                    (None, 64, 128, 2)   0           ['conv2d_transpose_1[0][0]',     
                                                                  'conv2d_4[0][0]']               
                                                                                                  
 conv2d_transpose_2 (Conv2DTran  (None, 512, 1024, 2  1026       ['add_1[0][0]']                  
 spose)                         )                                                                 
                                                                                                  
==================================================================================================
Total params: 139,581,132
Trainable params: 139,581,132
Non-trainable params: 0
```

### Source code / logs

**Custom Loss (common for both models)**

```
def self_balanced_focal_loss(alpha=3, gamma=2.0):
    def loss(y_true, y_pred):
        # cross entropy loss
        y_pred = backend.softmax(y_pred, -1)
        cross_entropy = backend.categorical_crossentropy(y_true, y_pred)

        # sample weights
        sample_weights = backend.max(backend.pow(1.0 - y_pred, gamma) * y_true, axis=-1)

        # class weights
        pixel_rate = backend.sum(y_true, axis=[1, 2], keepdims=True) / backend.sum(backend.ones_like(y_true),
                                                                                   axis=[1, 2], keepdims=True)
        class_weights = backend.max(backend.pow(backend.ones_like(y_true) * alpha, pixel_rate) * y_true, axis=-1)

        # final loss
        final_loss = class_weights * sample_weights * cross_entropy
        return backend.mean(backend.sum(final_loss, axis=[1, 2]))

    return loss
```
**GradCam Script (Common for both models)**

```
def get_img_array(img_path, size):   
    img = keras.preprocessing.image.load_img(img_path, target_size=size)
    array = keras.preprocessing.image.img_to_array(img)
    array = np.expand_dims(array, axis=0)
    return array


def make_gradcam_heatmap(img_array, model, last_conv_layer_name,idx1,idx2):
    # First, we create a model that maps the input image to the activations
    # of the last conv layer as well as the output predictions
    grad_model = tf.keras.models.Model(
        [model.inputs], [model.get_layer(last_conv_layer_name).output, model.output]
    )
   print(grad_model.outputs)
    # Then, we compute the gradient of the top predicted class for our input image
    # with respect to the activations of the last conv layer
    with tf.GradientTape() as tape:
        last_conv_layer_output, preds = grad_model(img_array)
        class_channel = preds[0][idx1][idx2][1]
        print("CLASS CHANNEL", class_channel)

    # This is the gradient of the output neuron (top predicted or chosen)
    # with regard to the output feature map of the last conv layer
    grads = tape.gradient(class_channel, last_conv_layer_output)
    print("GRADS", grads)


    # This is a vector where each entry is the mean intensity of the gradient
    # over a specific feature map channel
    pooled_grads = tf.reduce_mean(grads, axis=(0, 1, 2))

    # We multiply each channel in the feature map array
    # by "how important this channel is" with regard to the top predicted class
    # then sum all the channels to obtain the heatmap class activation
    last_conv_layer_output = last_conv_layer_output[0]
    heatmap = last_conv_layer_output @ pooled_grads[..., tf.newaxis]
    heatmap = tf.squeeze(heatmap)

    # For visualization purpose, we will also normalize the heatmap between 0 & 1
    heatmap = tf.maximum(heatmap, 0) / tf.math.reduce_max(heatmap)
    return heatmap.numpy()
```
Predictions from both the model yield useful results. However, the heatmap generated via the `gradcam` script for Model 2 returns 0's, and `anything/max(zeros)=NaN`

**In `Model 2`, there are no other changes, except minor ones, required to make the code compatible with TF>2, from TF==1.X. Why is it that the `tf.GradientTape()` is resulting in 0 values regardless of the image input, when the model (on the same images) is able to generate good predictions?**

Kindly advise and help,
Thanks.

Edit 01
Kindly check a difference I noticed in the GradCam results of the two models.
`Model 1` (Working Gradcam)
```
  [<KerasTensor: shape=(None, 16, 32, 512) dtype=float32 (created by layer 'block5_conv4')>, <KerasTensor: shape=(None, 256, 512, 2) dtype=float32 (created by layer 'conv2d_transpose_2')>]
CLASS CHANNEL tf.Tensor(0.26052475, shape=(), dtype=float32)
GRADS [-0.02966771 -0.02644881 -0.02611654 ...  0.01653964  0.01702563
  0.01976959]
```
`Model 2` (NaN/0 heatmap)
```
[<KerasTensor: shape=(None, 32, 64, 512) dtype=float32 (created by layer 'block5_conv4')>, <KerasTensor: shape=(None, 512, 1024, 2) dtype=float32 (created by layer 'conv2d_transpose_2')>]
CLASS CHANNEL tf.Tensor(-0.15865403, shape=(), dtype=float32)
GRADS [-0.00172401 -0.00150979 -0.00142616 ...  0.00059308  0.00060482
  0.00060655]
```
Thus, the major difference seems to be in the 'class_channel' variable ---> positive in Model 1.

**Comment by tilakrayal** [comment]
@varungupta31,
I ran the code shared and face a different error, please find the gist [here](https://colab.research.google.com/gist/tilakrayal/6d1689ac9b4c07526d15445cf26860b2/untitled415.ipynb) and share all dependencies to replicate the issue or share a colab gist with the reported error. Thank you!

**Comment by varungupta31** [comment]
@tilakrayal Hey! 
To run the code, you'll need to model file, which I can't share as of now.
Here is a complete [gist](https://colab.research.google.com/drive/1JJ6atTendiqCGjn8vMDpukz56m4qZ-b8?usp=sharing) containing relevant outputs and reported error (NaNs).

**Comment by tilakrayal** [comment]
@varungupta31,
I ran the code and faced a different error, please find the gist [here](https://colab.research.google.com/gist/tilakrayal/aef0d21138f331640b9035532eaebf6f/untitld99.ipynb) and share all dependencies to replicate the issue. Thank you!

**Comment by tilakrayal** [comment]
@varungupta31,
Without the required code, it would be difficult for us to debug the issue. In order to expedite the trouble-shooting process, could you please provide a complete code and the dependencies. Thank you!
